File: python/2055.蜡烛之间的盘子.py
class Solution:
    # [题解]1. O(n + m) t:324ms(63%) O(n) m:44.5M(28%)
    def platesBetweenCandles(self, s: str, queries: List[List[int]]) -> List[int]:
        n, m = len(s), len(queries)
        l, r, sum, ans = [0] * n, [0] * n, [0] * (n + 1), [0] * m
        p, q = -1, -1
        for i in range(n):
            if s[i] == '|': p = i
            if s[n - 1 - i] == '|': q = n - 1 - i
            l[i], r[n - i - 1] = p, q
            sum[i + 1] += sum[i] + (1 if s[i] == '*' else 0)
        for j in range(m):
            b, e = queries[j]
            c, d = r[b], l[e]
            if c != -1 and c <= d:
                ans[j] = sum[d + 1] - sum[c]
        return ans

File: cpp/2055.蜡烛之间的盘子.cpp
class Solution {
public:
    // [题解]1. O(m + n) t:372(73%) O(n) m:141.7M(27%) 前缀和
    vector<int> platesBetweenCandles(string s, vector<vector<int>>& queries) {
        int n = s.length();
        vector<int> l(n), r(n), sum(n + 1);
        for (int i = 0, j = n - 1, p = -1, q = -1; i < n; i++, j--){
            if (s[i] == '|') p = i;
            if (s[j] == '|') q = j;
            l[i] = p; r[j] = q;
            sum[i + 1] = sum[i] + (s[i] == '*' ? 1 : 0);
        }
        vector<int> ans;
        for (auto& query : queries){
            int a = query[0], b = query[1];
            int c = r[a], d = l[b];
            if (c != -1 && c <= d){
                ans.push_back(sum[d + 1] - sum[c]);
            }else{
                ans.push_back(0);
            }
        }
        return ans;
    }
};